sdk/core/azure-core/inc/azure/core/internal/deferred_operation.hpp

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+#include <exception>
+#include <functional>
+
+namespace Azure { namespace Core {
+  template <typename T> class DeferredResponse;
+
+  namespace _internal {
+
+    /**
+     * @brief Base class for deferred operation shared state. Wraps the outgoing request.
+     *
+     * This class exists to allow DeferredResponseSharedWithCallback objects to be aggregated
+     * where the specialization of the DeferredResponseSharedWithCallback has different
+     * specialization types.
+     *
+     */
+    class DeferredResponseSharedBase {
+      friend class DeferredResponseFactory;
+
+    public:
+      Azure::Core::Http::Request m_request;
+
+      virtual Azure::Core::Http::Request Request() { return m_request; }
+      /**
+       * @brief Process a raw response received from the service and save it for later
+       * retrieval.
+       *
+       * Called from the DeferredResponseProcessor when processing deferred operations.
+       *
+       * @param response Response returned by the service.
+       */
+      virtual void ProcessRawResponse(std::unique_ptr<Azure::Core::Http::RawResponse>& response)
+          = 0;
+
+      /**
+       * @brief Creates a new DeferredResponseSharedBase for the specified request.
+       *
+       * @param request Request to send to the server eventually.
+       */
+      explicit DeferredResponseSharedBase(Azure::Core::Http::Request request) : m_request(request)
+      {
+      }
+      virtual ~DeferredResponseSharedBase() {}
+    };
+
+    /**
+     * @brief Shared state for a DeferredResponse.
+     *
+     * This class implements a deferred operation For deferred operations which have a
+     * post-processing callback defined.
+     */
+    template <typename T> class DeferredResponseShared : public DeferredResponseSharedBase {
+    private:
+      std::function<Azure::Response<T>(std::unique_ptr<Azure::Core::Http::RawResponse>&)>
+          m_completeProcessing;
+      std::unique_ptr<Azure::Core::Http::RawResponse> m_rawResponse;
+
+    public:
+      explicit DeferredResponseShared(Azure::Core::Http::Request request)
+          : DeferredResponseShared(request, nullptr)
+      {
+      }
+      explicit DeferredResponseShared(
+          Azure::Core::Http::Request request,
+          std::function<Azure::Response<T>(std::unique_ptr<Azure::Core::Http::RawResponse>&)>
+              completeProcessing)
+          : DeferredResponseSharedBase(request), m_completeProcessing(completeProcessing)
+      {
+      }
+      virtual ~DeferredResponseShared() {}
+      void ProcessRawResponse(std::unique_ptr<Azure::Core::Http::RawResponse>& response)
+      {
+        m_rawResponse = std::move(response);
+      }
+
+      Response<T> GetResponse() { return m_completeProcessing(m_rawResponse); }
```

Comment:
This function is not declared as `virtual` so I assume it's not supposed to be overridden in derived class. In the function body, it unconditionally calls `m_completeProcessing`, which could be `nullptr` if the object was constructed by the first constructor. Then the application will be crashed. Is this by-design?

Reply:
It is not by design, it's a bug in the prototype. But the prototype is a proof of concept of the design and it's likely there are flaws in the implementation.

I can fix it if it's important, but this is a prototype.